app/src/main/java/io/appium/uiautomator2/utils/UiExpressionParser.java

```diff
@@ -0,0 +1,299 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.uiautomator2.utils;
+
+import android.support.annotation.Nullable;
+import android.support.test.uiautomator.UiObjectNotFoundException;
+import android.support.test.uiautomator.UiSelector;
+import android.util.Pair;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Stack;
+
+import io.appium.uiautomator2.common.exceptions.UiAutomator2Exception;
+import io.appium.uiautomator2.common.exceptions.UiSelectorSyntaxException;
+
+abstract class UiExpressionParser<T, U> {
+    protected final Class<T> clazz;
+    protected String expression;
+    private T target;
+
+    UiExpressionParser(Class<T> clazz, String expression) {
+        this.clazz = clazz;
+        this.expression = expression;
+        prepareForParsing();
+    }
+
+    protected String getConstructorExpression() {
+        return "new " + clazz.getSimpleName();
+    }
+
+    public abstract U parse() throws UiSelectorSyntaxException, UiObjectNotFoundException;
+
+    // prepares text for the main parsing loop
+    protected void prepareForParsing() {
+        expression = expression.trim();
+        if (expression.startsWith(clazz.getSimpleName())) {
+            expression = "new " + expression;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected void consumeConstructor() throws UiSelectorSyntaxException,
+            UiObjectNotFoundException {
+        expression = expression.trim();
+        if (!expression.startsWith(getConstructorExpression())) {
+            throw new UiSelectorSyntaxException(String.format(
+                    "Was trying to parse as %1$s, but didn't start with an acceptable prefix. " +
+                            "Acceptable prefixes are: 'new %1$s' or '%1$s'. Saw:%2$s",
+                    clazz.getSimpleName(), expression));
+        }
+        expression = expression.substring(getConstructorExpression().length());
+        final List<String> params = consumeMethodParameters();
+        final Pair<Constructor, List<Object>> constructor = findConstructor(params);
+        try {
+            target = (T) constructor.first.newInstance(constructor.second.toArray());
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new UiSelectorSyntaxException("Can not create instance of " +
+                    clazz.getSimpleName(), e);
+        }
+    }
+
+    protected void consumePeriod() throws UiSelectorSyntaxException {
+        expression = expression.trim();
+        if (expression.startsWith(".")) {
+            expression = expression.substring(1);
```

Comment:
should we verify expression length first?

Reply:
No, we shouldn't - the string length is always >=1 if ```expression.startsWith(".") == true```

Reply:
how about just a single dot `"."`?

Reply:
The result of ```"x".substring(1)``` is empty string.